### PR TITLE
[DDP] Support for multiple backwards

### PR DESCRIPTION
**Pull request by rohan-varma**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#59359 [DDP] Support for multiple backwards**

Move `prepare_for_backward` into `_DDPSink` backward instead of calling it in DDP forward pass so that we can run multiple backwards in DDP with `retain_graph=True`. 

ci-all version of this PR: https://github.com/pytorch/pytorch/pull/59931

Differential Revision: [D28855226](https://our.internmc.facebook.com/intern/diff/D28855226/)

### CODE REVIEW

File: torch/nn/parallel/distributed.py [hunk 1]
```diff
@@ -840,37 +845,30 @@ def forward(self, *inputs, **kwargs):
...
+            self.require_forward_param_sync = (
+                grad_enabled and self.require_backward_grad_sync
             )
-            # Reconstruct output data structure.
-            output = tree_unflatten(passthrough_tensor_list, treespec)
+
+        state_dict = {
+            'static_graph': self.static_graph,
+            'find_unused': self.find_unused_parameters,
+            'num_iterations': self.num_iterations,
```
Comment (zhaojuanmao) [inline torch/nn/parallel/distributed.py:855]: self.num_iterations is calculated for forward pass in python side, may not be synced with num_bwd_calls_ in C++

File: torch/nn/parallel/distributed.py [hunk 2]
```diff
@@ -130,18 +130,23 @@ def backward(ctx, *grad_outputs):
         if static_graph_training and ctx.state_dict['num_iterations'] == 1:
             Variable._execution_engine.queue_callback(ctx.reducer._delay_all_reduce)
```
Comment (zhaojuanmao) [inline torch/nn/parallel/distributed.py:131]: this may need to be moved after prepare_for_backward() call, and use ctx.reduer.get_num_backward_calls_() to check whether it is first backward iteration

**Review by zhaojuanmao — COMMENTED** [review]
overall looks good! left one comment

File: torch/nn/parallel/distributed.py (hunk 1 above)
Comment (rohan-varma) [reply torch/nn/parallel/distributed.py:855]: Agreed it is not synced, but we really only use this to decide whether we need to enqueue delay allreduce right? And we only enqueue delay allreduce exactly once (basically only the first call to backward), so just checking num_iterations == 1 should be sufficient as delay allreduce will run in the next bwd.

Although agreed it might be a bit cleaner to directly check the num_backward_calls, perhaps

File: torch/nn/parallel/distributed.py (hunk 2 above)
Comment (rohan-varma) [reply torch/nn/parallel/distributed.py:131]: Would it be simpler if we just have a boolean flag in python associated with static graph training? We always only need to enqueue it once, and only the first time, so something like `enqueued_delay_allreduce` flag should be enough and simpler here?

Comment (rohan-varma) [reply torch/nn/parallel/distributed.py:131]: Also if there are N forward calls before a backward, checking `num_backward_calls` in from C++ may result in us enqueueing this multiple times, which is not wanted

Comment (rohan-varma) [reply torch/nn/parallel/distributed.py]: discussed offline, we are using  ctx.reduer.get_num_backward_calls_() as suggested

File: torch/nn/parallel/distributed.py (hunk 1 above)
Comment (rohan-varma) [reply torch/nn/parallel/distributed.py:860]: discussed offline, we are using  ctx.reduer.get_num_backward_calls_() as suggested

**Review by zhaojuanmao — APPROVED** [review]
looks great, thanks for adding this!

File: torch/lib/c10d/reducer.cpp [hunk 3]
```diff
@@ -1486,6 +1495,13 @@ void Reducer::finalize_backward() {
     local_used_maps_reduced_ = false;
   }
 
+  // Reset bucket counting to ensure we can appropriately launch allreduce for
+  // each bucket in the next backwards.
+  reset_bucket_counting();
+  // If we populated rebuilt params list in this backward call, avoid
+  // repopulating in subsequent backward calls.
+  all_rebuilt_params_pushed_ = !rebuilt_params_.empty();
```
Comment (zhaojuanmao) [inline torch/lib/c10d/reducer.cpp:1505]: rebuilt_params_() will be cleared after rebuild_bucket(), in this case, all_rebuilt_params_pushed_ will be true in every iteration after rebuilding bucket, and turns out to push rebuilt params many times, right?

File: torch/lib/c10d/reducer.cpp [hunk 4]
```diff
@@ -1486,6 +1495,13 @@ void Reducer::finalize_backward() {
     local_used_maps_reduced_ = false;
   }
 
+  // Reset bucket counting to ensure we can appropriately launch allreduce for
+  // each bucket in the next backwards.
+  reset_bucket_counting();
```
Comment (zhaojuanmao) [inline torch/lib/c10d/reducer.cpp:1500]: why this needs to be moved out of prepare_backward() and called at multiple places instead?

File: torch/lib/c10d/reducer.cpp (hunk 3 above)
Comment (rohan-varma) [reply torch/lib/c10d/reducer.cpp:1505]: Good point, although pushing rebuilt params is also guarded by `should_rebuild_bucket()` call, which will return false after we have rebuilt buckets, so this wouldn't cause an issue from what I can tell.

Although, it's probably better to change this to:
```
all_rebuilt_params_pushed = all_rebuilt_params_pushed || !rebuilt_params_.empty()

``

File: torch/lib/c10d/reducer.cpp (hunk 4 above)
Comment (rohan-varma) [reply torch/lib/c10d/reducer.cpp]: Discussed offline, added a helper function to call this at finalize_backward() and rebuild_buckets() where it is required.

File: torch/nn/parallel/distributed.py [hunk 5]
```diff
@@ -142,23 +142,35 @@ def forward(ctx, reducer, state_dict, *inputs):
...
+                # with undefined gradient to maintain parity with local training.
+                # DDP takes care of undefined grads in this case to ensure the .grad
+                # field of the param is not touched.
+                ctx.reducer.prepare_for_backward(list(_find_tensors(ctx.inputs)))
+
+        # Note that we enqueue delay allreduce after prepare_for_backward in
+        # static graph training as prepare_for_backward sets the
+        # num_backwards_call counter in the reducer.
+        static_graph_first_bwd = (
+            static_graph_training and ctx.reducer._static_graph_first_bwd()
```
Comment (zhaojuanmao) [inline torch/nn/parallel/distributed.py:169]: nit: I'm wondering why static_graph_training is needed here? _static_graph_first_bwd() has already checked this?

**Review by zhaojuanmao — COMMENTED** [review]
looks good! just one minor question

File: torch/nn/parallel/distributed.py (hunk 5 above)
Comment (rohan-varma) [reply torch/nn/parallel/distributed.py:169]: If we are not in static graph mode at all, it would save one Python -> C++ -> Python round trip. I feel like that might be worth it as a small optimization in non-static graph mode.

**Comment by rohan-varma** [comment]
Ci errors are unrelated, landing
